client/src/homepage/viewed-documents.tsx

```diff
@@ -0,0 +1,94 @@
+import React, { useEffect, useState } from "react";
+import { Link } from "react-router-dom";
+
+import { usePageVisibility } from "./hooks";
+
+type Entry = {
+  url: string;
+  title: string;
+  timestamp: number;
+};
+
+// Simpler and cheaper than a proper library
+function friendlyDateDisplay(date: Date): string {
+  const today = new Date();
+  const dateString = date.toDateString();
+  const secondsDiff = (today.getTime() - date.getTime()) / 1000;
+  if (secondsDiff < 60) {
+    return "seconds ago";
+  }
+  if (secondsDiff < 60 * 15) {
+    return "minutes ago";
+  }
+  if (today.toDateString() === dateString) {
+    return "today";
+  }
+  const yesterday = new Date(today.getTime() - 1000 * 3600 * 24);
+  if (yesterday.toDateString() === dateString) {
+    return "yesterday";
+  }
+  return dateString;
+}
+
+export default function ViewedDocuments() {
+  const isVisible = usePageVisibility();
+  const [entries, setEntries] = useState<Entry[] | null>(null);
+
+  useEffect(() => {
+    if (isVisible) {
+      const localStorageKey = "viewed-documents";
+      const previousVisits = JSON.parse(
+        localStorage.getItem(localStorageKey) || "[]"
+      );
+      const newEntries: Entry[] = [];
+      // console.log(previousVisits);
+      for (const visit of previousVisits) {
+        newEntries.push({
+          url: visit.url,
+          title: visit.title,
+          timestamp: visit.timestamp,
+        });
+      }
+
+      if (newEntries.length) {
+        setEntries(newEntries);
+      }
+    }
+  }, [isVisible]);
+
+  return (
+    <div>
+      <h3>Recently viewed documents</h3>
```

Comment:
```suggestion
      <h2>Recently viewed documents</h2>
```
Perhaps an `h2` here? We can also go next-level and mark this up for additional semantics as follows:

```suggestion
  <article aria-labelledby="recently-viewed-documents">
      <h2 id="recently-viewed-documents">Recently viewed documents</h2>
```

an `article` being "a complete, or self-contained, composition in a document, page, application, or site and that is, in principle, independently distributable or reusable"

My vote would be for the second suggestion.

Reply:
Should the `<h2>Welcome to MDN</h2>` then become `<h1>Welcome to MDN</h1>`?

Reply:
Indeed, it should.